lib/internal/modules/esm/hooks.js

```diff
@@ -150,13 +151,16 @@ class Hooks {
     } = pluckHooks(exports);
 
     if (globalPreload) {
-      ArrayPrototypePush(this.#chains.globalPreload, { fn: globalPreload, url });
+      const next = this.#chains.globalPreload[this.#chains.globalPreload.length - 1];
+      ArrayPrototypePush(this.#chains.globalPreload, { fn: globalPreload, url, next });
     }
     if (resolve) {
-      ArrayPrototypePush(this.#chains.resolve, { fn: resolve, url });
+      const next = this.#chains.resolve[this.#chains.resolve.length - 1];
+      ArrayPrototypePush(this.#chains.resolve, { fn: resolve, url, next });
```

Comment:
What if there’s only one loader in the chain? So if `length` was 1, `next` would be set to the same function as `this.#chains.resolve`, rather than `defaultResolve`?

Reply:
I'm very certain this is safe currently… there's just sets the `next` entry to be the last entry currently in the chain list, essentially forming a reversed singly-linked list. If there's no entries in the list when this is invoked, then `next` is `undefined`, which is fine because that represents the end of the chain per https://github.com/nodejs/node/pull/48559/files#diff-15847fb1d42ec31618a186ad743b13da05f05a0708f93399f7cba82088ac8a8dR750.

> So if length was 1, next would be set to the same function as this.#chains.resolve, rather than defaultResolve? 

If the length is`1` that means there is no custom loaders currently present – chain link index 0 is always `defaultResolve`:

```js
  chains = {
    // ...
    resolve: [
      {
        fn: defaultResolve,
        url: 'node:internal/modules/esm/resolve',
      },
    ],
    // ...
  }
```